peps/pep-0758.rst

```diff
@@ -0,0 +1,432 @@
+  the headers and their values as described above.
+
+Changes to tools
+================
+
+Theoretically, installer tools shouldn't need any changes, since when they
+have identified the wheel to download and install, they simply consult the
+download URLs returned by PyPI's Simple API. In practice though, tools such as
+``pip`` and ``uv`` may have constrained lists of hosts they will allow
+downloads from, such as PyPI's own ``pythonhosted.org`` domain.
+
+In this case, such tools will need to relax those constraints, but the exact
+policy for this is left to the installer tools themselves. Any number of
+approaches could be implemented, such as downloading the ``.rim`` file and
+verifying the ``EXTERNAL-HOSTING`` metadata, or simply trusting the external
+downloads for any wheel with a matching checksum.  They could also query PyPI
+for the project's organization owner, then verify that it has a support URI
+before trusting the download. They could warn the user when externally hosted
+wheel files are encountered, and/or require the use of a command line option
+to enable additional download hosts. Any of these verification policies could
+be chosen in configuration files.
+
+Installer tools should also probably provide better error messages when
+externally hosted wheels cannot be downloaded, e.g. because a host is
+unreachable. As described above, such tools could query enough metadata from
+PyPI to provide clear and distinct error messages pointing users to the
+package's external hosting support email or issue tracker.
+
+Security
+========
+
+Several factors as described in this proposal should mitigate security
+concerns with externally hosted wheels, such as:
+
+- Wheel file checksums MUST be included in ``.rim`` files, and once uploaded
+  cannot be changed. Since the checksum stored on PyPI is immutable and
+  required, it is not possible to spoof an external wheel file, even if the
+  owning organization lost control of their hosting domain.
+- Externally hosted wheels MUST be served over HTTPS.
+- In order to serve externally hosted wheels, organizations MUST be approved
+  by the PyPI admins.
+
+When users identify malware or vulnerabilities in PyPI-hosted projects, they
+can now report this using the `malware reporting facilities
+<https://pypi.org/security/>`__ on PyPI, as also described in this
+`blog post
+<https://blog.pypi.org/posts/2024-03-06-malware-reporting-evolved/>`__.  The
+same process can be used to report security issues in externally hosted
+wheels, and the same remediation process should be used.  In addition, since
+organizations with external hosting enabled MUST provide a support contact
+URI, that URI can be used in some cases to report the security issue to the
+hosting organization.  Such organization reporting won't make sense for
+malware, but could indeed be a very useful way to report security
+vulnerabilities in externally hosted wheels.
+
```

Comment:
I think it would be useful to add a section listing requirements for external hosts in technical respects:

- MUST be served over HTTPS signed by a certificate trusted by Mozilla's root certificate store (which is what pip uses to verify certificates)
- MUST support HTTP range requests. This is used by some installers as a fallback if they cannot read the `.metadata` files PyPi serves, and I don't think this should degrade.
- Maybe should require HTTP/2 for multiplexing connections? Unsure if that is needed or useful.
- SHOULD use a CDN (orgs will want to do this anyway, but this is a good idea to recommend, even if not required)
- MUST commit to a stable URL for all wheel packages
- CAN remove an external file ONLY if that file is deleted from PyPi first. MUST NOT do so if yanked. This allows for handling IP removals/legal issues.

Reply:
I think it does make sense to be explicit about external host requirements.

> MUST be served over HTTPS signed by a certificate trusted by Mozilla's root certificate store (which is what pip uses to verify certificates)

It looks like pip and uv are somewhat different in the details.
* [pip](https://pip.pypa.io/en/stable/topics/https-certificates/) by default (as of 24.2 and Python 3.10) uses the system cert store in addition to the [certifi](https://pypi.org/project/certifi/) (i.e. Mozilla) store.
* [uv](https://docs.astral.sh/uv/configuration/authentication/#custom-ca-certificates) uses the (Mozilla) `webpki-roots` store by default, but *not* the system store unless you include the `--native-tls` flag, although it's not clear from the docs whether this adds the system root certs to Mozilla's or replaces them.

So I guess it looks like the Mozilla one is the common denominator and probably should be required, but I feel like I want to hedge a little bit.  Hmm.  Maybe "MUST be served over HTTPs which SHOULD be trusted by Mozilla's root certificate, but MUST at least be compatible with pip's and uv's default stores"

Reply:
> MUST support HTTP range requests. This is used by some installers as a fallback if they cannot read the .metadata files PyPi serves, and I don't think this should degrade.

Do you have any references for that?  I checked the pip and uv documentation, and the [PEP 658](https://peps.python.org/pep-0658/) docs, but couldn't find any discussion of this.  Do you know what installers support this behavior?

Reply:
> Maybe should require HTTP/2 for multiplexing connections? Unsure if that is needed or useful.

Similarly, do you know of any installers that use this?  I couldn't find any mention of this in the pip or uv documentation.

Reply:
The last three bullet items in your list sound like good recommendations.

Reply:
I am a little worried about being too loose in defining what certs need to be supported. I think this may need to be something that PyPi checks when setting up an external host. I think it would be best to say the certs should be trusted by Mozilla's cert bundle, but the PyPi admins may modify the list of acceptable certificates. This allows flexibility for future modifications but makes the initial acceptable set extremely clear (well, as clear as certificate verification can be).

Reply:
Also it looks like uv *replaces* the Mozilla bundle with the system cert store when using `--native-tls`. If you are doing that however it is up to you to verify that you are compatible with the certificates for the software you download at the end of the day.

Reply:
Re: uv; yep, confirmed with Charlie that it [replaces](https://github.com/astral-sh/uv/blob/3ce34035c84804fdfb8b78cf11b9ba1b168d0f35/crates/uv-client/src/base_client.rs#L248).

Reply:
> I am a little worried about being too loose in defining what certs need to be supported. I think this may need to be something that PyPi checks when setting up an external host. 

I suspect that in practice, support for Mozilla will always be the case, given that you want to *already* be compatible with pip and uv, but I'll come up with something that addresses your concerns.  Thanks!

Reply:
> Do you have any references for that? I checked the pip and uv documentation, and the [PEP 658](https://peps.python.org/pep-0658/) docs, but couldn't find any discussion of this. Do you know what installers support this behavior?

I know for sure at least pip does (ref https://github.com/pypa/pip/blob/102d8187a1f5a4cd5de7a549fd8a9af34e89a54f/src/pip/_internal/network/lazy_wheel.py#L26-L29), and I just checked that uv does as well: https://github.com/astral-sh/uv/blob/ed1684a0e4caee45482098666da124b5aa195b5c/crates/uv-client/src/registry_client.rs#L600-L609

It isn't currently a requirement for a simple index, but PyPi supports it and I don't think that should regress because of external hosting. As a practical matter, supporting HTTP range requests is very easy if you are doing anything reasonable to serve wheels, such as putting them on S3, putting them behind a load balancer like nginx, etc.

> Maybe should require HTTP/2 for multiplexing connections? Unsure if that is needed or useful.
>
>Similarly, do you know of any installers that use this? I couldn't find any mention of this in the pip or uv documentation.

HTTP/2 is also a feature PyPi supports, and I think it would be bad to regress. While I am not sure if pip or uv currently use it, being able to multiplex connections could help with parallel downloading of wheels, which I know pip is exploring.

I am not knowledgeable enough about pip and uv to confidently say whether or not this is required, but HTTP/2 does provide some nice benefits, and I think it should be considered to be required. I don't think it is as important as range requests.

Reply:
Thanks for the links.  Are you comfortable with making both of these **SHOULD** requirements?

Reply:
I think range requests is a MUST, since it is such an important optimization for large wheels, but HTTP/2 is fine to be SHOULD. I was looking at a pip install log and it seems like it was making HTTP/1.1 requests against PyPi, so HTTP/2 may be feasible to just not include in the requirements.

Reply:
I think the details of SHOULD/MUST can be figured out on discuss.python.org.

Reply:
Okay, sounds good.  I'll make ranges MUST and HTTP/2 SHOULD.

I think the thing that makes me a little uncomfortable about this section (to be pushed soon) is that these are requirements on external wheel hosts that PyPI can't really enforce.  The wording I'm trying to come up with is something like: `The following constraints lead to reliable and compatible external wheel hosting services`.

Let's see how that goes over when this gets posted to DPO.